### PR TITLE
Result store update

### PR DESCRIPTION
**Pull request by DanTanAtAims**
Update result store to use YAXArrays and NetCDF. 

After each counterfactual and intervention run call.
```julia
append_all_results(result_store::ResultStore, start_year::Int, end_year::Int, reps::Int)
```
This will add the results to the YAXArray dataset and add rows to the scenario DataFrame reflecting scenario intervention factors.  

```bash
├─RME_outcomes_yyyy-mm-dd-HH-MM-SS
│       results.nc
│       scenarios.csv
```

### CODE REVIEW

File: src/ReefModEngine.jl [hunk 1]
```diff
@@ -23,6 +23,17 @@ macro RME(func)
     end
 end
 
+"""
+Only for use when RME functions return non-error numeric results. 
+
+For example any getter retrieving a numberic property.
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:29]: `numberic` -> `numeric`

File: src/ReefModEngine.jl [hunk 2]
```diff
@@ -23,6 +23,17 @@ macro RME(func)
...
+
+ivOutplantCountPerM2(::Cstring)::Cdouble.
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:31]: Put inside a an `# Examples` section please with properly fenced code section

````
# Examples

```julia
ivOutplantCountPerM2(::Cstring)::Cdouble
```
````

File: src/ReefModEngine.jl [hunk 3]
```diff
@@ -23,6 +23,17 @@ macro RME(func)
...
+"""
+macro getRME(func)
+    return esc(Meta.parse("@ccall RME.$(func)"))
+end
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:35]: Wow, how does this work? I thought you needed to explicitly specify what the return type was for `@ccall`?

File: src/ReefModEngine.jl [hunk 4]
```diff
@@ -44,6 +55,6 @@ export
 
 # IO
 export
-    ResultStore, collect_all_results!, collect_rep_results!, save_to_mat
+    ResultStore, append_all_results!, save_to_mat
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:58]: If append is actually concat, should we call it `concat_results` instead?

File: src/ReefModEngine.jl (hunk 3 above)
Comment (DanTanAtAims) [reply src/ReefModEngine.jl:35]: Unfortunately, we still need to specify the return type. 
```julia
count::Float64 = @getRME ivOutplantCountPerM2("<name>"::Cstring)::Cdouble
```
Most of the calls to the RME use an out variable and the return value is the integer returned. 
This error integer is checked in the `@RME` macro.
```julia
!(typeof(err) <: Cstring) && err != 0 && !(typeof(err) <: Cvoid)
```
For functions that return numeric types the condition `err != 0` causes an exception when nothing has gone wrong. 

The documentation is somewhat unclear about how these functions 'error' when the return value is not an error indicator.

Comment (ConnectedSystems) [reply src/ReefModEngine.jl:35]: Oh I get it, the return type is part of the `func` being parsed.

Thanks!

Comment (ConnectedSystems) [reply src/ReefModEngine.jl:35]: > For functions that return numeric types the condition err != 0 causes an exception when nothing has gone wrong.

Oh, I forgot about that. Yes it's a pain that 0 stands for both "an error occurred" and "no error occurred" 😢

File: src/ReefModEngine.jl [hunk 5]
```diff
@@ -26,9 +26,11 @@ end
 """
 Only for use when RME functions return non-error numeric results. 
 
-For example any getter retrieving a numberic property.
+# Examples
 
-ivOutplantCountPerM2(::Cstring)::Cdouble.
+```julia
+count::Float64 = @getRME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble.
+```
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:33]: Hey @DanTanAtAims what's the difference between the two below?

```
@getRME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble

# and

@RME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble
```

File: src/ReefModEngine.jl [hunk 6]
```diff
@@ -26,9 +26,11 @@ end
 """
 Only for use when RME functions return non-error numeric results. 
 
-For example any getter retrieving a numberic property.
+# Examples
 
-ivOutplantCountPerM2(::Cstring)::Cdouble.
+```julia
+count::Float64 = @getRME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble.
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:32]: Delete full stop

```suggestion
count::Float64 = @getRME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble
```

File: src/ReefModEngine.jl (hunk 5 above)
Comment (DanTanAtAims) [reply src/ReefModEngine.jl:33]: The second will error out when there was no error. In the RME macro 
```julia
macro RME(func)
    local m = esc(Meta.parse("@ccall RME.$(func)"))
    return quote
        local err = $m

---->   if !(typeof(err) <: Cstring) && err != 0 && !(typeof(err) <: Cvoid) # <--- this condition prevents function calls that return numbers to fail.
            local err_msg = @ccall RME.lastError()::Cstring
            throw(ArgumentError("Call to RME failed with message: $(unsafe_string(err_msg))"))
        end
        ...
    end
end
```
The conditions of the if statement are evaluated as follows for getters that return numeric, non-error values
```julia
!(Cdouble <: Cstring) (true)
err != 0 (true) # Importantly no error has occurred. 
!(typeof(err) <: Cvoid) (true)
```
So we enter the error `if` block. We should only use the new macro `@getRME` when we have no choice as we lose the safety of the error checking in the original macro.

Comment (ConnectedSystems) [reply src/ReefModEngine.jl]: Yep, obvious now, sorry.

File: docs/src/getting_started.md [hunk 7]
```diff
@@ -84,13 +84,10 @@ match_id("10-330")
...
 # Create a convenient result store to help extract data from RME
-result_store = ResultStore(start_year, end_year, reps)
-
-# Collect results for runs 1, 3 and 5 only
-collect_rep_results!(result_store, start_year, end_year, [1,3,5])
+result_store = ResultStore(start_year, end_year)
 
 # Collect and store all results, where `reps` is the total number of expected runs.
-collect_all_results!(result_store, start_year, end_year, reps)
+append_all_results!(result_store, start_year, end_year, reps)
```
Comment (ConnectedSystems) [inline docs/src/getting_started.md:90]: These examples need to be updated

**Review by ConnectedSystems — COMMENTED** [review]
Very last comment to address Dan, thank you.

File: src/ReefModEngine.jl [hunk 8]
```diff
@@ -23,6 +23,19 @@ macro RME(func)
     end
 end
 
+"""
+Only for use when RME functions return non-error numeric results. 
+
+# Examples
+
+```julia
+count::Float64 = @getRME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble
```
Comment (ConnectedSystems) [inline src/ReefModEngine.jl:32]: One more - update example variable as `count` is also a function.

```suggestion
count_per_m2::Float64 = @getRME ivOutplantCountPerM2("iv_name"::Cstring)::Cdouble
```

**Review by ConnectedSystems — APPROVED** [review]
Thanks @DanTanAtAims

**Pull request merged**